### ISSUE TITLE
Reuse session tokens

### ISSUE THREAD

**Issue by mconigliaro** [zcutlip/pyonepassword#21]
According to `op signin --help`:

```
You can use the `--session` option with the `signin` command to reuse an
active token or to test whether a session has expired. This may be
useful when writing scripts that use the command-line tool. If the
session is active, the tool will use it and return the same token. If
the session is expired, you’ll have to sign in again.
```

Currently, `pyonepassword.OP()` prompts me for a password every time, despite the fact that I've already logged in, and `OP_SESSION_<shorthand>` is set in my environment.

**Comment by zcutlip** [comment]
I'll look into this, it bugs me too.

I *think* I thought through this at the time and had some reason I did it the way I did, but worth revisiting for sure. Thanks for the issue

**Comment by mconigliaro** [comment]
It looks like you already have the logic to detect the shorthand for the last login. I think you just need to look for the existence of the `OP_SESSION_<shorthand>` env var and only sign in when `op` requires it. It looks like `op` says "You are not currently signed in" any time you run a command with an expired token, so maybe you only sign in when that happens?

**Comment by zcutlip** [comment]
I have a solution if you want to try it out on the `dev/reuse_session` branch.

After igging into it I see why I didn't attempt to automatically reuse sessions. It *mostly* comes down to it being hard to distinguish between `op`'s failure modes, e.g., a wrong password vs session expiration vs a "get item" query failing or some other reason entirely. `op` just exits with `1` in all cases, and I didn't want to scrape stderr, which would be brittle and unreliable (and probably not work with non-English localizations). 

I also wanted `op` to fall back to an interactive prompt if the caller doesn't want to:

- Handle prompting for and handling the password themselves, or
- trust `pyonepassword` to handle their password safely

That said, the original behavior is still default. But you can instantiate `OP` like so, if you want it to reuse a session:

```python
op = OP(use_existing_session=True)
```

If you don't want an interactive password prompt, and instead want an exception, pass `password_prompt=False`:
```python
try:
  op = OP(use_existing_session=True, password_prompt=False)
except OPNotSignedInException:
 #  handle getting a password for signing in
  op = OP(password=password)
```

Additionally, if you want to save your session environment variable and token, `op` now has two properties to retrieve the token value and the session variable name, respectively:

```python
op.token
op.session_var
```

You can use these values to set/export the session environment variable in whatever way you choose, so they can be reused outside the current python process, e.g., if you run your script again.

**Comment by mconigliaro** [comment]
Yes, I also noticed that `op` only seems to use two exit codes (0 and 1), so I created [this thread](https://1password.community/discussion/123649/exit-code-documentation/) for you.

In general, I think it could be helpful to move the authentication stuff to its own method. That way, you could instantiate new `OP` objects without authenticating automatically and just call the `signin` command on demand (e.g. either on the command line or something like `OP.signin()` in your library). 

But this looks like a great step in the right direction. I don't know when I'm going to have time to check this out, but I'll be back! 😎

**Comment by mconigliaro** [comment]
Yes, your intent makes sense. I was just thinking it could be useful to allow calls to `signin` on demand, in the same spirit as allowing users to prompt for the password on their own. The `OP` constructor could just use that same `signin` function, or people could use `OP(signin=False)` to call it themselves. Just a thought.

**Comment by zcutlip** [comment]
Hi, sorry for taking so long on this.

I don't object to exposing a `signin()` method on the `OP` object, but I'm not clear how it's useful.

Are you thinking of instantiating an `OP` without signing in (e.g., in hopes of reusing a session) and then if it raises an exception, call the signin() method? If so, is this better than just instantiating a new `OP` object with sign-in credentials? They don't maintain any state (in terms of querying data from your vault), so discarding it and creating a new one shouldn't lose any information.

Again, totally open to it, but just want to understand the use case a bit.

**Comment by zcutlip** [comment]
Got it; misunderstood. Thank you!
